feat: add new cryptography subject for building basic zkp protocol from scratch (#2)
quartz-technology/workshops: cryptography/building_a_simple_zkp_protocol_from_scratch/README.md

# Build a simple ZKP Protocol from scratch

> In this workshop, you will create a very simple Zero-Knowledge Proof protocol that can be used by a

Author: Added in 6b961c700511ddb991df0bb610c29f860646a336

### ✅ Validation

Print the two polynomials to verify their content. Here is the snippet of code to achieve this:
```python
print(f'Target Polynomial: {target_polynomial}')
```

Which should print, when executed with `python3 common.py`:
```shell
Target Polynomial: 2.0 - 3.0·x¹ + 1.0·x²
```

### 📚 Documentation

- [NumPy documentation on how to declare polynomials](https://numpy.org/doc/stable/reference/routines.polynomials.html#polynomials)

## Step 2: Verifier Test Data Sampling and Prover Evaluation

### 📑 Description

Now that the target polynomial can be used by the verifier, he must be able to sample test data used
by the prover for a direct evaluation.

### 📌 Tasks

1. Create a `verifier.py` file.
   1. Generate a random number `r`. Use the `random.randint(-127, 128)` method
   call to sample a random `int8`.
   2. Evaluate `t = int(target_polynomial(r))` using the target polynomial found in the `common.py`
   file.
   3. Print the variable `r`.
2. In the `prover.py` file:
   1. After the last line, add an `input()` call which stores the value passed in the standard input
   in a variable called `r`. Don't forget to surround the input call with an integer cast `int(..)`.

>    2. Compute the value of a variable called `p` equal to the evaluation of the secret polynomial

Reviewer: ```suggestion
   2. Compute the value of a variable `p` equal to the evaluation of the secret polynomial
```

   using the value `r`.

>    3. Compute the value of a variable called `h` equal to the evaluation of the arbitrary

Reviewer: ```suggestion
   3. Compute the value of a variable `h` equal to the evaluation of the arbitrary
```

   polynomial using the value `r`.

>    4. Print the values `p` and `h`.

Reviewer: ```suggestion
   4. Print the variables `p` and `h`.
```

### ✅ Validation

In a first terminal, run the following command:
```shell
python3 verifier.py
```

> Which should print (note that the data sampled can differ as it is a random number):

Reviewer: ```suggestion
Which should print (note that the data sampled can differ as it depends on a randomly generated number):
```

Test Data Sampled: -46
```

Copy the `Test Data Sampled` value and in a second terminal, run the following command:
```shell
python3 prover.py
```

Nothing should be printed yet. Paste the copied value and hit enter. You should get the following (
the output should also differ due to randomness):
```shell
Evaluation of p(x): -103776.0
Evaluation of h(x): -46.0
```

### 📚 Documentation

- [NumPy documentation on how to use polynomials](https://numpy.org/doc/stable/reference/routines.polynomials.html#polynomials)

## Step 3: Verifier's Verification

### 📑 Description

In this last step, the verifier will use the value evaluated by the prover and its own evaluated
value to test if the condition `p = t * h` holds.

### 📌 Tasks

1. In the `verifier.py` file:
   1. Add two `int(float(input()))` calls and store the result in two variables, namely `p` and `h`.
   2. Compute `h * t`.
   3. If the resulting value is equal to `p`, print the following sentence: `Okay, it seems you
   know all the roots for your third degree polynomial`.
   4. Otherwise, print: `Mmmmh, I think you don't know all the roots of this third degree
   polynomial`

### ✅ Validation

> Repeat the process from the last step.

Reviewer: ```suggestion
Repeat the process from step 2.
```

After printing the values on the prover's side, copy and paste the first one in the verifier's
terminal and repeat this for the second value evaluated by the prover.

If you've copied the values properly, you should get the valid message.
But for example, if you pasted different values, which represents a prover that does not know `p(x)`
and samples an invalid value, the condition won't hold on the verifier's side and the invalid message
should be printed to the screen. Let's try this:

Verifier's run:
```shell
Target Polynomial: 2.0 - 3.0·x¹ + 1.0·x²
Test Data Sampled: 59
```

Now a valid prover should evaluate `p` and `h` as `195054.0` and `59` respectively.
If you voluntarily choose different values, for example `12` and `44` and paste those in the verifier's shell
who's waiting for us, you should get:
```shell
12
44
Mmmmh, I think you don't know all the roots of this third degree polynomial
```

### 📚 Documentation

> Nothing here !

Reviewer: Then remove it

Author: Just a typo, thks !

Author: Removed in 1aec933fa45548320dbc71bf14c758d45fcb44db

## Possible Improvements

> This is a very basic implementation. You can improve it for example by not having the polynomials

Reviewer: What about a secured version? What about doing the same for a 4-degree polynomial?

Author: This is the topic of the next subjects 👀
